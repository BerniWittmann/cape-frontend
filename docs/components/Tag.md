# Tag

A Component to view a single Tag

## Props

<!-- @vuese:Tag:props:start -->
|Name|Description|Type|Required|Default|
|---|---|---|---|---|
|tag|The Tag object|`Object`|`true`|-|
|closable|Option whether the tag should be deleteable or not|`Boolean`|`false`|false|
|size|Size of the Tag|`'medium'` / `'small'` / `'mini'`|`false`|-|

<!-- @vuese:Tag:props:end -->


## Events

<!-- @vuese:Tag:events:start -->
|Event Name|Description|Parameters|
|---|---|---|
|close|Emitted when the close button is clicked|-|

<!-- @vuese:Tag:events:end -->


