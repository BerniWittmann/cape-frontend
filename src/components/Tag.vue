<template>
  <el-tag class="tag" type="info" :style="style" :closable="closable" :size="size" @close="emitClose">{{ tag.name }}</el-tag>
</template>

<script>
import { convertHexToRgba } from '@/utils/helpers'

/*
 * @vuese
 * @group Components
 *
 * A Component to view a single Tag
 */
export default {
  name: 'Tag',
  props: {
    // The Tag object
    tag: {
      type: Object,
      required: true
    },

    // Option whether the tag should be deleteable or not
    closable: {
      type: Boolean,
      default: false
    },

    // Size of the Tag
    size: {
      // `'medium'` / `'small'` / `'mini'`
      type: String
    }
  },

  computed: {
    style() {
      if (!this.tag.color) return {}

      return {
        backgroundColor: convertHexToRgba(this.tag.color, 0.1),
        borderColor: convertHexToRgba(this.tag.color, 0.2),
        color: convertHexToRgba(this.tag.color, 1)
      }
    }
  },

  methods: {
    emitClose() {
      // Emitted when the close button is clicked
      this.$emit('close')
    }
  }
}

</script>

<style lang="scss">
.tag {
  margin-right: 5px;
  &:last-of-type {
    margin-right: 0;
  }

  .el-tag__close {
    color: inherit !important;
    &:hover {
      background-color: inherit !important;
      color: #fff;
    }
  }
}
</style>
