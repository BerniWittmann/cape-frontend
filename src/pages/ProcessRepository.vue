<template>
  <v-layout>
    <el-row justify="end" :gutter="20">
      <el-col :span="6" :offset="18">
        <el-input
                :placeholder="$t('process.search')"
                v-model="search">
          <i slot="prefix" class="el-input__icon el-icon-search"></i>
        </el-input>
      </el-col>
    </el-row>
    <el-row :gutter="20">
      <el-col :span="24">
        <el-table
                :data="processes"
                :default-sort="{prop: 'lastEditedAt', order: 'descending'}">
          <el-table-column
                  prop="name"
                  :label="$t('process.name')"
                  sortable>
          </el-table-column>
          <el-table-column
                  prop="createdAt"
                  :label="$t('process.created_at')"
                  sortable>
            <template slot-scope="scope">
              {{ formatDateTime(scope.row.createdAt) }}
            </template>
          </el-table-column>
          <el-table-column
                  prop="lastEditedAt"
                  :label="$t('process.last_edited_at')"
                  sortable>
            <template slot-scope="scope">
              {{ formatDateTime(scope.row.lastEditedAt) }}
            </template>
          </el-table-column>
          <el-table-column
                  prop="tags"
                  :label="$t('process.tags')"
                  :filters="availableTags"
                  :filter-method="filterTag"
                  filter-placement="bottom-end">
            <template slot-scope="scope">
              <tag v-for="tag in scope.row.tags" :key="tag.id" :tag="tag" class="row-tag"></tag>
            </template>
          </el-table-column>
          <el-table-column
                  width="60">
            <template slot-scope="scope">
              <el-tooltip :content="$t('process.edit.tooltip')" placement="top">
              <el-button
                      @click.native.prevent="edit(scope.row)"
                      circle size="mini"
                      icon="el-icon-edit"></el-button>
              </el-tooltip>
            </template>
          </el-table-column>
        </el-table>
      </el-col>
    </el-row>
  </v-layout>
</template>

<script>
/* ============
 * Process Repository Page
 * ============
 *
 * The Process Repository Page
 */
import { DATE_TIME_FORMAT } from '@/utils/constants'
import DefaultLayout from '@/layouts/Default.vue'
import Tag from '@/components/Tag.vue'

export default {
  components: {
    VLayout: DefaultLayout,
    Tag
  },

  data() {
    return {
      search: ''
    }
  },

  computed: {
    allProcesses() {
      return this.$store.state.process.processes
    },

    processes() {
      return this.allProcesses.filter(p => {
        if (!this.search || this.search.length === 0) return true

        return p.name.toLowerCase().includes(this.search.toLowerCase())
      })
    },

    availableTags() {
      let tags = {}

      this.allProcesses.forEach(p => {
        p.tags.forEach(t => {
          if (!tags[t.id]) {
            tags[t.id] = { value: t.id, text: t.name }
          }
        })
      })

      return Object.values(tags)
    }
  },

  methods: {
    formatDateTime(datetimeObj) {
      return datetimeObj.format(DATE_TIME_FORMAT)
    },

    filterTag(value, row) {
      return row.tags.find(t => t.id === value)
    },

    edit(process) {
      this.$router.push({
        name: 'process.edit',
        params: {
          processID: process.id
        }
      })
    }
  }
}
</script>

<style scoped lang="scss">

</style>
