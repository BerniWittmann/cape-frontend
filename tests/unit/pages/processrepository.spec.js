import { shallowMount, mount } from '@vue/test-utils'
import { i18n } from '../setupPlugins'
import moment from 'moment'
import ProcessRepository from '@/pages/ProcessRepository.vue'
import Tag from '@/components/Tag.vue'
import EmptySlotComponent from '../EmptySlotComponent.vue'

describe('Pages', () => {
  describe('ProcessRepository.vue', () => {
    let store
    let cmp

    const date = moment(moment().utc())

    beforeEach(() => {
      store = {
        state: {
          process: {
            processes: [{
              id: '1',
              name: 'First Process',
              createdAt: date.clone().subtract(14, 'days'),
              lastEditedAt: date.clone().subtract(2, 'days'),
              tags: [{
                id: '42',
                name: 'First Tag',
                color: '#FF0000'
              }, {
                id: '43',
                name: 'Second Tag',
                color: '#FFFF00'
              }]
            }, {
              id: '2',
              name: 'Second Process',
              createdAt: date.clone().subtract(8, 'days'),
              lastEditedAt: date.clone().subtract(6, 'days'),
              tags: [{
                id: '42',
                name: 'First Tag',
                color: '#FF0000'
              }]
            }, {
              id: '3',
              name: 'Third Process',
              createdAt: date.clone().subtract(2, 'days'),
              lastEditedAt: date.clone().subtract(1, 'days'),
              tags: []
            }]
          }
        }
      }
      render()
    })

    function render() {
      cmp = shallowMount(ProcessRepository, {
        i18n,
        mocks: {
          $store: store
        },
        stubs: {
          'v-layout': EmptySlotComponent
        }
      })
    }

    it('renders', () => {
      expect(cmp.html()).toMatchSnapshot()
    })

    it('renders the table correctly', (done) => {
      // Hide error resulting from el-table warning
      console.error = jest.fn()

      cmp = mount(ProcessRepository, {
        i18n,
        mocks: {
          $store: store
        },
        stubs: {
          'v-layout': EmptySlotComponent
        },
        sync: false
      })
      cmp.vm.$nextTick(() => {
        expect(cmp.html()).toMatchSnapshot()
        done()
      })
    })

    it('renders the table', () => {
      expect(cmp.contains('eltable-stub')).toBeTruthy()
    })

    it('renders all processes', () => {
      const table = cmp.find('eltable-stub')

      expect(table.props('data')).toEqual(store.state.process.processes)
    })

    it('can search for a process', () => {
      const searchInput = cmp.find('elinput-stub')

      expect(searchInput.exists()).toBeTruthy()

      cmp.vm.search = 'sec'

      const table = cmp.find('eltable-stub')

      expect(table.props('data')).toEqual([store.state.process.processes[1]])
    })

    it('can filter for process tags', () => {
      const tagColumn = cmp.findAll('eltablecolumn-stub').at(3)
      expect(tagColumn.exists()).toBeTruthy()

      expect(tagColumn.props('filters')).toEqual([{
        value: '42',
        text: 'First Tag'
      }, {
        value: '43',
        text: 'Second Tag'
      }])
    })

    it('can filter for a tag', () => {
      expect(cmp.vm.filterTag('42', store.state.process.processes[0])).toEqual({
        color: '#FF0000',
        id: '42',
        name: 'First Tag'
      })
    })

    it('renders the tag components', (done) => {
      cmp = mount(ProcessRepository, {
        i18n,
        mocks: {
          $store: store
        },
        stubs: {
          'v-layout': EmptySlotComponent
        },
        sync: false
      })
      cmp.vm.$nextTick(() => {
        expect(cmp.contains(Tag)).toBeTruthy()
        done()
      })
    })
  })
})
