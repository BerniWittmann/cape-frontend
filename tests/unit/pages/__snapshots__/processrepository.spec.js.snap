// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pages ProcessRepository.vue renders 1`] = `
<div>
  <elrow-stub tag="div" gutter="20" justify="end" align="top">
    <elcol-stub span="6" tag="div" offset="18">
      <elinput-stub value="" type="text" autocomplete="off" validateevent="true" placeholder="process.search"><i class="el-input__icon el-icon-search"></i></elinput-stub>
    </elcol-stub>
  </elrow-stub>
  <elrow-stub tag="div" gutter="20" justify="start" align="top">
    <elcol-stub span="24" tag="div">
      <eltable-stub data="[object Object],[object Object],[object Object]" fit="true" showheader="true" defaultsort="[object Object]" selectonindeterminate="true">
        <eltablecolumn-stub type="default" label="process.name" prop="name" sortable="" resizable="true" filtermultiple="true" sortorders="ascending,descending,"></eltablecolumn-stub>
        <eltablecolumn-stub type="default" label="process.created_at" prop="createdAt" sortable="" resizable="true" filtermultiple="true" sortorders="ascending,descending,"></eltablecolumn-stub>
        <eltablecolumn-stub type="default" label="process.last_edited_at" prop="lastEditedAt" sortable="" resizable="true" filtermultiple="true" sortorders="ascending,descending,"></eltablecolumn-stub>
        <eltablecolumn-stub type="default" label="process.tags" prop="tags" resizable="true" filtermethod="function () { [native code] }" filters="[object Object],[object Object]" filterplacement="bottom-end" filtermultiple="true" sortorders="ascending,descending,"></eltablecolumn-stub>
      </eltable-stub>
    </elcol-stub>
  </elrow-stub>
</div>
`;

exports[`Pages ProcessRepository.vue renders the table correctly 1`] = `
<div>
  <div class="el-row is-justify-end" style="margin-left: -10px; margin-right: -10px;">
    <div class="el-col el-col-6 el-col-offset-18" style="padding-left: 10px; padding-right: 10px;">
      <div class="el-input el-input--prefix">
        <!----><input type="text" autocomplete="off" placeholder="process.search" class="el-input__inner"><span class="el-input__prefix"><i class="el-input__icon el-icon-search"></i><!----></span>
        <!---->
        <!---->
      </div>
    </div>
  </div>
  <div class="el-row" style="margin-left: -10px; margin-right: -10px;">
    <div class="el-col el-col-24" style="padding-left: 10px; padding-right: 10px;">
      <div class="el-table el-table--fit el-table--scrollable-x el-table--enable-row-hover el-table--enable-row-transition">
        <div class="hidden-columns">
          <div></div>
          <div></div>
          <div></div>
          <div></div>
        </div>
        <div class="el-table__header-wrapper">
          <table cellspacing="0" cellpadding="0" border="0" class="el-table__header" style="width: 320px;">
            <colgroup>
              <col name="el-table_1_column_1" width="80">
              <col name="el-table_1_column_2" width="80">
              <col name="el-table_1_column_3" width="80">
              <col name="el-table_1_column_4" width="80">
            </colgroup>
            <thead class="">
              <tr class="">
                <th colspan="1" rowspan="1" class="el-table_1_column_1     is-leaf is-sortable">
                  <div class="cell">process.name<span class="caret-wrapper"><i class="sort-caret ascending"></i><i class="sort-caret descending"></i></span></div>
                </th>
                <th colspan="1" rowspan="1" class="el-table_1_column_2     is-leaf is-sortable">
                  <div class="cell">process.created_at<span class="caret-wrapper"><i class="sort-caret ascending"></i><i class="sort-caret descending"></i></span></div>
                </th>
                <th colspan="1" rowspan="1" class="el-table_1_column_3     is-leaf is-sortable">
                  <div class="cell">process.last_edited_at<span class="caret-wrapper"><i class="sort-caret ascending"></i><i class="sort-caret descending"></i></span></div>
                </th>
                <th colspan="1" rowspan="1" class="el-table_1_column_4     is-leaf">
                  <div class="cell">process.tags<span class="el-table__column-filter-trigger"><i class="el-icon-arrow-down"></i></span></div>
                </th>
              </tr>
            </thead>
          </table>
        </div>
        <div class="el-table__body-wrapper is-scrolling-left">
          <table cellspacing="0" cellpadding="0" border="0" class="el-table__body" style="width: 320px;">
            <colgroup>
              <col name="el-table_1_column_1" width="80">
              <col name="el-table_1_column_2" width="80">
              <col name="el-table_1_column_3" width="80">
              <col name="el-table_1_column_4" width="80">
            </colgroup>
            <tbody>
              <tr class="el-table__row">
                <td rowspan="1" colspan="1" class="el-table_1_column_1  ">
                  <div class="cell">First Process</div>
                </td>
                <td rowspan="1" colspan="1" class="el-table_1_column_2  ">
                  <div class="cell">
                    05.11.2018 21:26
                  </div>
                </td>
                <td rowspan="1" colspan="1" class="el-table_1_column_3  ">
                  <div class="cell">
                    17.11.2018 21:26
                  </div>
                </td>
                <td rowspan="1" colspan="1" class="el-table_1_column_4  ">
                  <div class="cell"><span class="tag row-tag el-tag el-tag--info" style="background-color: rgba(255, 0, 0, 0.1); color: rgb(255, 0, 0);" name="el-zoom-in-center">First Tag</span><span class="tag row-tag el-tag el-tag--info" style="background-color: rgba(255, 255, 0, 0.1); color: rgb(255, 255, 0);" name="el-zoom-in-center">Second Tag</span></div>
                </td>
              </tr>
              <tr class="el-table__row">
                <td rowspan="1" colspan="1" class="el-table_1_column_1  ">
                  <div class="cell">Second Process</div>
                </td>
                <td rowspan="1" colspan="1" class="el-table_1_column_2  ">
                  <div class="cell">
                    11.11.2018 21:26
                  </div>
                </td>
                <td rowspan="1" colspan="1" class="el-table_1_column_3  ">
                  <div class="cell">
                    13.11.2018 21:26
                  </div>
                </td>
                <td rowspan="1" colspan="1" class="el-table_1_column_4  ">
                  <div class="cell"><span class="tag row-tag el-tag el-tag--info" style="background-color: rgba(255, 0, 0, 0.1); color: rgb(255, 0, 0);" name="el-zoom-in-center">First Tag</span></div>
                </td>
              </tr>
              <tr class="el-table__row">
                <td rowspan="1" colspan="1" class="el-table_1_column_1  ">
                  <div class="cell">Third Process</div>
                </td>
                <td rowspan="1" colspan="1" class="el-table_1_column_2  ">
                  <div class="cell">
                    17.11.2018 21:26
                  </div>
                </td>
                <td rowspan="1" colspan="1" class="el-table_1_column_3  ">
                  <div class="cell">
                    18.11.2018 21:26
                  </div>
                </td>
                <td rowspan="1" colspan="1" class="el-table_1_column_4  ">
                  <div class="cell"></div>
                </td>
              </tr>
              <!---->
            </tbody>
          </table>
          <!---->
          <!---->
        </div>
        <!---->
        <!---->
        <!---->
        <!---->
        <div class="el-table__column-resize-proxy" style="display: none;"></div>
      </div>
    </div>
  </div>
</div>
`;
